lib/db/lowlevel.go

```diff
@@ -7,88 +7,93 @@
 package db
 
 import (
-	"os"
-	"strings"
-	"sync"
+	"bytes"
+	"encoding/binary"
+	"fmt"
 	"sync/atomic"
 
-	"github.com/syndtr/goleveldb/leveldb"
-	"github.com/syndtr/goleveldb/leveldb/errors"
 	"github.com/syndtr/goleveldb/leveldb/iterator"
-	"github.com/syndtr/goleveldb/leveldb/opt"
-	"github.com/syndtr/goleveldb/leveldb/storage"
 	"github.com/syndtr/goleveldb/leveldb/util"
-)
 
-const (
-	dbMaxOpenFiles = 100
-	dbWriteBuffer  = 16 << 20
-	dbFlushBatch   = dbWriteBuffer / 4 // Some leeway for any leveldb in-memory optimizations
+	"github.com/syncthing/syncthing/lib/protocol"
 )
 
+// Backend is the interface that a database needs to implement to be used with
+// Syncthing.
+// It is not guaranteed that after calling Close no more calls are made to other
+// methods (see ErrorChecker.IsClosed).
+type Backend interface {
+	ErrorChecker
+	Get(key []byte) ([]byte, error)
+	Has(key []byte) (bool, error)
+	Put(key, val []byte) error
+	Delete(key []byte) error
+	NewBatch() Batch
+	NewIterator(slice *util.Range) (iterator.Iterator, error)
+	GetSnapshot() (Snapshot, error)
+	Close()
+}
+
+// IsClosed should return true if the given error is a result of the Backend
+// already being closed.
+//
+// IsClosed should return true if the given error is a result of a key not
+// being found in the database.
+type ErrorChecker interface {
+	IsClosed(err error) bool
+	IsNotFound(err error) bool
+}
+
+// Batch is a container to check in several changes at once to the database to
+// ensure a semantically consistent state.
+//
+// CheckFlush should check the size of the Batch write and call Flush in
+// appropriate intervals (e.g. when the Batch exceeds a certain size).
+//
+// Flush commits the contents of the Batch to db and calls Reset on it on success.
+type Batch interface {
+	CheckFlush() error
```

Comment:
My reasoning is, that it can depend on the db when a good size to flush is reached. Letting the db flush on `Put` if necessary would invalidate the purpose of `Batch`: Collecting changes together that should be done atomically (semantically). I.e. `CheckFlush` is caller saying to the db, that it may flush at this point if the db thinks it's a good idea, and `Flush` says do it now (likely because operation ended and nothing more is being added to Batch).

I could create an `flushBatch` type used by `Interface` that takes care of flushing based on size regardless of the underlying `Backend`, then `CheckFlush` could be removed from the interface.